code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstanceHostMapAgentProcessHandler.java

```diff
@@ -0,0 +1,108 @@
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.eventing.model.Event;
+import io.cattle.platform.eventing.model.EventVO;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.object.util.DataUtils;
+import io.cattle.platform.process.common.handler.AgentBasedProcessHandler;
+import io.cattle.platform.util.type.CollectionUtils;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class InstanceHostMapAgentProcessHandler extends AgentBasedProcessHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(InstanceHostMapAgentProcessHandler.class);
+
+    private static final String CLUSTER_CONNECTION_FIELD = "clusterConnection";
+
+    @Inject
+    ClusterManager clusterManager;
+
+    @Inject
+    ClusterHostMapDao clusterHostMapDao;
+
+    @Inject
+    IpAddressDao ipAddressDao;
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        // TODO: Hack dataResource
+        if (state.getResource() instanceof InstanceHostMapRecord) {
+            // Check if host is a cluster.  If so, augment event with cluster connection info so that agent
+            // can connect to cluster server instead of regular docker daemon.
+            InstanceHostMapRecord instanceHostMap = (InstanceHostMapRecord) state.getResource();
+            Long hostId = instanceHostMap.getHostId();
+            Host potentialCluster = objectManager.loadResource(Host.class, hostId);
+            if (objectManager.isKind(potentialCluster, ClusterConstants.KIND)) {
+                Long managingHostId = DataAccessor.fields(potentialCluster).withKey(ClusterConstants.MANAGING_HOST).as(Long.class);
+                Integer clusterServerPort = DataAccessor.fields(potentialCluster).withKey(ClusterConstants.CLUSTER_SERVER_PORT).as(Integer.class);
+
+                Instance clusterServerInstance = clusterManager.getClusterServerInstance(potentialCluster);
+                IpAddress ipAddress = clusterManager.getClusterServerInstanceIp(clusterServerInstance);
+
+                DataUtils.getWritableFields(instanceHostMap).put(CLUSTER_CONNECTION_FIELD, "http://" + ipAddress.getAddress() + ":" + clusterServerPort);
+
+                instanceHostMap.setHostId(managingHostId);
+            }
+        }
+        return super.handle(state, process);
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    protected void postProcessEvent(EventVO<?> event, Event reply, ProcessState state, ProcessInstance process,
+            Object eventResource, Object dataResource, Object agentResource) {
+        Host actualHost = null;
+
+        List names = (List)CollectionUtils.getNestedValue(reply.getData(),
```

Comment:
An alternative to the host name solution might be to always add the host uuid to the response event. It should always uniquely identify the host.

AgentResourceMonitorImpl makes use of it.
https://github.com/rancherio/cattle/blob/c552571c3d2003446cc5af9269563cc4da3258db/code/iaas/agent-server/src/main/java/io/cattle/platform/agent/server/resource/impl/AgentResourcesMonitorImpl.java

NOTE: host uuid and physical host uuid are two different things. You would not want physcial host.